pkg/policy/selectorcache.go

```diff
@@ -590,9 +595,27 @@ func (sc *SelectorCache) AddFQDNSelector(user CachedSelectionUser, fqdnSelec api
 	// current selections after adding a selector. This way the
 	// behavior is the same between the two cases here (selector
 	// is already cached, or is a new one).
+
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
+
+	// Check whether the selectorCache was updated while 'newFQDNSel' was
+	// being registered without the 'sc.mutex'. If so, use it. Otherwise
+	// we can use the one we just created/configured above.
+	if sel := sc.lookupSelectorLocked(key); sel != nil {
+		// Unregister, discard our local copy.
+		newFQDNSel.dnsProxy.UnregisterForIdentityUpdates(newFQDNSel.selector)
```

Comment:
Remove this line - we should only unregister if there are no more users, and if there is an entry in the `SelectorCache` for this FQDNSelector, it will have users.

Reply:
Thanks for the review, the `RegisterForIdentityUpdates()` function has a different set of side-effects than I thought and this forced some rethinking of the way we patch this up. I've fixed this up in the latest version.